File: frontend/src/services/emergencyLocation.ts
/**
 * Emergency Location Service for Autonomous Accident Detection & AI-DR (SIH26168).
 * 
 * Determines the best available navigation state when an accident occurs:
 * - GPS HEALTHY  -> Uses existing GPS / fused navigation position (position_source = 'GPS')
 * - GPS DEGRADED -> Uses existing best sensor-fusion/AI-DR position (position_source = 'SENSOR_FUSION')
 * - GPS LOST     -> Uses existing AI-DR estimated position (position_source = 'AI_DR')
 * 
 * CRITICAL ARCHITECTURE RULE:
 * - Must NOT call browser GPS (navigator.geolocation).
 * - Must NOT create an independent location calculation.
 * - Must reuse the location already calculated by the existing AI-DR navigation engine.
 */

import { EmergencyLocation, PositionSource, SIHFrame } from '../types';

const API_BASE_URL =
  import.meta.env.VITE_API_URL || "http://127.0.0.1:8001";

/**
 * Pure extraction function that determines the best emergency location from an active navigation state.
 * Never touches browser GPS or creates a duplicate location calculation.
 */
export function getEmergencyLocation(
  navState?: Partial<SIHFrame> | Record<string, any> | null
): EmergencyLocation {
  // Baseline Delhi reference coordinates if no state provided
  const state: any = navState || {
    latitude: 28.6139,
    longitude: 77.2090,
    altitude: 216.0,
    speed: 0.0,
    speed_kmh: 0.0,
    heading: 0.0,
    gps_status: 'HEALTHY',
    ai_confidence_pct: 98.0,
    uncertainty_sigma_m: 1.5,
    timestamp: Date.now() / 1000
  };

  // 1. Identify GPS Status
  const rawStatus = (
    state.gps_status ||
    state.gps_health?.health_state ||
    state.timeline_stage ||
    ''
  ).toString().toUpperCase();

  const isOutage = Boolean(state.is_outage || state.is_outage_active);

  let normalizedStatus: 'HEALTHY' | 'DEGRADED' | 'LOST' = 'HEALTHY';
  if (isOutage || rawStatus.includes('LOST') || rawStatus.includes('DENIED') || rawStatus.includes('OUTAGE') || rawStatus.includes('AI-DR')) {
    normalizedStatus = 'LOST';
  } else if (rawStatus.includes('DEGRADED') || rawStatus.includes('NOISY') || rawStatus.includes('UNRELIABLE')) {
    normalizedStatus = 'DEGRADED';
  } else {
    normalizedStatus = 'HEALTHY';
  }

  // 2. Extract Common Telemetry Attributes
  const speed = Number(state.speed_kmh ?? ((state.speed ?? 0) * 3.6));
  const heading = Number(state.heading ?? state.fused_heading ?? 0);
  const altitude = state.altitude !== undefined ? Number(state.altitude) : (state.estimated_altitude !== undefined ? Number(state.estimated_altitude) : undefined);
  const timestamp = state.timestamp ?? Date.now() / 1000;

  const sensorTrust = state.sensor_trust || {};
  const gpsTrust = Number(sensorTrust.gps_reliability ?? 99.0);
  const imuTrust = Number(sensorTrust.imu_reliability ?? 95.0);
  const motionTrust = Number(sensorTrust.motion_reliability ?? 92.0);

  // 3. Select Position Strictly According to GPS Health State
  let positionSource: PositionSource = 'GPS';
  let latitude: number;
  let longitude: number;
  let confidence: number;
  let estimatedErrorM: number;
  let navReliability: number;

  if (normalizedStatus === 'LOST') {
    // -------------------------------------------------------------
    // GPS LOST -> Strictly use existing AI-DR estimated position
    // -------------------------------------------------------------
    positionSource = 'AI_DR';

    if (state.aidr_latitude !== undefined && state.aidr_latitude !== null) {
      latitude = Number(state.aidr_latitude);
      longitude = Number(state.aidr_longitude);
    } else if (state.dr_latitude !== undefined && state.dr_latitude !== null) {
      latitude = Number(state.dr_latitude);
      longitude = Number(state.dr_longitude);
    } else if (state.estimated_latitude !== undefined && state.estimated_latitude !== null) {
      latitude = Number(state.estimated_latitude);
      longitude = Number(state.estimated_longitude);
    } else {
      latitude = Number(state.latitude ?? 0);
      longitude = Number(state.longitude ?? 0);
    }

    confidence = Number(state.confidence ?? state.ai_confidence_pct ?? state.confidence_pct ?? 86.0);
    estimatedErrorM = Number(state.estimated_error_m ?? state.aidr_error_m ?? state.uncertainty_sigma_m ?? state.dr_error_m ?? 15.0);
    navReliability = Number(((imuTrust + motionTrust) / 2).toFixed(1));

  } else if (normalizedStatus === 'DEGRADED') {
    // -------------------------------------------------------------
    // GPS DEGRADED -> Use existing Sensor Fusion / AI-DR position
    // -------------------------------------------------------------
    if (state.smooth_latitude !== undefined && state.smooth_latitude !== null) {
      latitude = Number(state.smooth_latitude);
      longitude = Number(state.smooth_longitude);
      positionSource = 'SENSOR_FUSION';
    } else if (state.fused_latitude !== undefined && state.fused_latitude !== null) {
      latitude = Number(state.fused_latitude);
      longitude = Number(state.fused_longitude);
      positionSource = 'SENSOR_FUSION';
    } else if (state.aidr_latitude !== undefined && state.aidr_latitude !== null) {
      latitude = Number(state.aidr_latitude);
      longitude = Number(state.aidr_longitude);
      positionSource = 'AI_DR';
    } else {
      latitude = Number(state.latitude ?? 0);
      longitude = Number(state.longitude ?? 0);
      positionSource = 'SENSOR_FUSION';
    }

    confidence = Number(state.confidence ?? state.ai_confidence_pct ?? 74.0);
    estimatedErrorM = Number(state.estimated_error_m ?? state.smooth_error_m ?? state.uncertainty_sigma_m ?? 5.2);
    navReliability = Number((gpsTrust * 0.4 + imuTrust * 0.6).toFixed(1));

  } else {
    // -------------------------------------------------------------
    // GPS HEALTHY / AVAILABLE -> Use existing GPS fix
    // -------------------------------------------------------------
    positionSource = 'GPS';
    latitude = Number(state.latitude ?? state.smooth_latitude ?? 0);
    longitude = Number(state.longitude ?? state.smooth_longitude ?? 0);

    confidence = Number(state.confidence ?? state.ai_confidence_pct ?? 98.0);
    estimatedErrorM = Number(state.estimated_error_m ?? state.uncertainty_sigma_m ?? 1.5);
    navReliability = Number(gpsTrust.toFixed(1));
  }

  return {
    latitude: Number(latitude.toFixed(7)),
    longitude: Number(longitude.toFixed(7)),
    altitude,
    speed: Number(speed.toFixed(1)),
    heading: Number(heading.toFixed(1)),
    gps_status: normalizedStatus,
    position_source: positionSource,
    confidence: Number(confidence.toFixed(1)),
    estimated_error_m: Number(estimatedErrorM.toFixed(2)),
    navigation_reliability: navReliability,
    timestamp
  };
}

/**
 * Fetch current emergency location from backend API
 */
export async function fetchEmergencyLocation(gpsStatus?: string): Promise<EmergencyLocation> {
  const url = new URL(`${API_BASE_URL}/emergency/location`);
  if (gpsStatus) {
    url.searchParams.append('gps_status', gpsStatus);
  }
  const res = await fetch(url.toString());
  if (!res.ok) throw new Error('Failed to fetch emergency location');
  return res.json();
}

File: navigation/emergency_location.py
"""
Emergency Location Provider for Autonomous Accident Detection & AI-DR (SIH26168).
Extracts the best available navigation state without creating redundant location systems.

Selects position source strictly according to GPS signal integrity:
- GPS HEALTHY  -> Uses existing GPS / Fused GNSS position (position_source = 'GPS').
- GPS DEGRADED -> Uses existing Sensor Fusion / AI-DR position (position_source = 'SENSOR_FUSION').
- GPS LOST     -> Uses existing AI-DR Machine Learning estimated position (position_source = 'AI_DR').

CRITICAL SAFETY & SYSTEM ARCHITECTURE RULE:
- NEVER invokes browser GPS (navigator.geolocation).
- NEVER creates an independent or duplicate location calculation.
- Directly reuses the location already calculated by the existing AI-DR navigation engine.
"""

import time
from typing import Dict, Any, Optional


class NavigationStateTracker:
    """
    In-memory registry of the current active navigation state.
    Updated as the navigation pipeline (SIH demo, EKF fusion, AI-DR corrector, or sensor stream) executes.
    """
    _current_nav_state: Optional[Dict[str, Any]] = None

    @classmethod
    def set_current_state(cls, state: Dict[str, Any]) -> None:
        """Update active navigation state."""
        cls._current_nav_state = state

    @classmethod
    def get_current_state(cls) -> Optional[Dict[str, Any]]:
        """Retrieve active navigation state."""
        return cls._current_nav_state

    @classmethod
    def clear(cls) -> None:
        """Clear active navigation state."""
        cls._current_nav_state = None


def get_emergency_location(nav_state: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """
    Determines and returns the best available vehicle position from the existing navigation state.
    
    Parameters:
        nav_state: Optional dictionary containing navigation frame data.
                   If None, retrieves the latest active state from NavigationStateTracker.
                   
    Returns:
        Dict with keys:
            - latitude: float
            - longitude: float
            - altitude: float (if available)
            - speed: float
            - heading: float
            - gps_status: str ('HEALTHY', 'DEGRADED', 'LOST')
            - position_source: str ('GPS', 'SENSOR_FUSION', 'AI_DR')
            - confidence: float (0..100)
            - estimated_error_m: float
            - navigation_reliability: float (0..100)
            - timestamp: float
    """
    # 1. Resolve source navigation state
    if nav_state is None:
        nav_state = NavigationStateTracker.get_current_state()

    # Fallback to nominal Delhi reference coordinates if no state has been registered yet
    if nav_state is None:
        nav_state = {
            "latitude": 28.6139,
            "longitude": 77.2090,
            "altitude": 216.0,
            "speed": 0.0,
            "heading": 0.0,
            "gps_status": "HEALTHY",
            "ai_confidence_pct": 98.0,
            "uncertainty_sigma_m": 1.5,
            "sensor_trust": {"gps_reliability": 99.0, "imu_reliability": 97.0, "motion_reliability": 95.0},
            "timestamp": round(time.time(), 2)
        }

    # 2. Extract and normalize GPS status
    raw_status = (
        nav_state.get("gps_status") or
        nav_state.get("gps_health", {}).get("health_state") or
        nav_state.get("timeline_stage") or
        ""
    )
    is_outage = bool(nav_state.get("is_outage", False) or nav_state.get("is_outage_active", False))

    status_str = str(raw_status).upper()
    if is_outage or "LOST" in status_str or "DENIED" in status_str or "OUTAGE" in status_str or "AI-DR" in status_str:
        normalized_status = "LOST"
    elif "DEGRADED" in status_str or "NOISY" in status_str or "UNRELIABLE" in status_str:
        normalized_status = "DEGRADED"
    else:
        normalized_status = "HEALTHY"

    # 3. Extract core telemetry attributes
    t_curr = float(nav_state.get("timestamp", round(time.time(), 2)))
    
    # Speed in km/h
    if "speed_kmh" in nav_state:
        speed_val = round(float(nav_state["speed_kmh"]), 1)
    elif "speed" in nav_state:
        speed_val = round(float(nav_state["speed"]) * 3.6, 1)
    else:
        speed_val = round(float(nav_state.get("velocity", 0.0)) * 3.6, 1)

    heading_val = round(float(nav_state.get("heading", nav_state.get("fused_heading", 0.0))), 1)

    altitude_val = nav_state.get("altitude")
    if altitude_val is None:
        altitude_val = nav_state.get("estimated_altitude", nav_state.get("fused_altitude"))
    if altitude_val is not None:
        altitude_val = round(float(altitude_val), 2)

    # Extract sensor trust & reliability
    sensor_trust = nav_state.get("sensor_trust", {})
    gps_trust = float(sensor_trust.get("gps_reliability", 99.0))
    imu_trust = float(sensor_trust.get("imu_reliability", 95.0))
    motion_trust = float(sensor_trust.get("motion_reliability", 92.0))

    # 4. Determine best available location strictly according to GPS health
    if normalized_status == "LOST":
        # -------------------------------------------------------------
        # GPS LOST -> Use existing AI-DR estimated position
        # CRITICAL: Do NOT use browser GPS. Do NOT use raw/lost GPS.
        # -------------------------------------------------------------
        position_source = "AI_DR"
        gps_status_out = "LOST"

        # AI-DR estimated position priority
        if "aidr_latitude" in nav_state and nav_state["aidr_latitude"] is not None:
            lat = round(float(nav_state["aidr_latitude"]), 7)
            lon = round(float(nav_state["aidr_longitude"]), 7)
        elif "dr_latitude" in nav_state and nav_state["dr_latitude"] is not None:
            lat = round(float(nav_state["dr_latitude"]), 7)
            lon = round(float(nav_state["dr_longitude"]), 7)
        elif "estimated_latitude" in nav_state and nav_state["estimated_latitude"] is not None:
            lat = round(float(nav_state["estimated_latitude"]), 7)
            lon = round(float(nav_state["estimated_longitude"]), 7)
        else:
            lat = round(float(nav_state.get("latitude", 0.0)), 7)
            lon = round(float(nav_state.get("longitude", 0.0)), 7)

        # AI-DR confidence during outage
        confidence_val = nav_state.get("confidence") or nav_state.get("ai_confidence_pct") or nav_state.get("confidence_pct", 86.0)
        confidence = round(float(confidence_val), 1)

        # Error bounds from ML predictor or uncertainty sigma
        error_val = (
            nav_state.get("estimated_error_m") or
            nav_state.get("aidr_error_m") or
            nav_state.get("uncertainty_sigma_m") or
            nav_state.get("dr_error_m") or
            15.0
        )
        estimated_error = round(float(error_val), 2)

        # Navigation reliability based on active IMU & Motion models
        nav_reliability = round((imu_trust + motion_trust) / 2.0, 1)

    elif normalized_status == "DEGRADED":
        # -------------------------------------------------------------
        # GPS DEGRADED -> Use best sensor-fusion / AI-DR position
        # -------------------------------------------------------------
        gps_status_out = "DEGRADED"

        if "smooth_latitude" in nav_state and nav_state["smooth_latitude"] is not None:
            lat = round(float(nav_state["smooth_latitude"]), 7)
            lon = round(float(nav_state["smooth_longitude"]), 7)
            position_source = "SENSOR_FUSION"
        elif "fused_latitude" in nav_state and nav_state["fused_latitude"] is not None:
            lat = round(float(nav_state["fused_latitude"]), 7)
            lon = round(float(nav_state["fused_longitude"]), 7)
            position_source = "SENSOR_FUSION"
        elif "aidr_latitude" in nav_state and nav_state["aidr_latitude"] is not None:
            lat = round(float(nav_state["aidr_latitude"]), 7)
            lon = round(float(nav_state["aidr_longitude"]), 7)
            position_source = "AI_DR"
        else:
            lat = round(float(nav_state.get("latitude", 0.0)), 7)
            lon = round(float(nav_state.get("longitude", 0.0)), 7)
            position_source = "SENSOR_FUSION"

        confidence_val = nav_state.get("confidence") or nav_state.get("ai_confidence_pct") or 72.0
        confidence = round(float(confidence_val), 1)

        error_val = (
            nav_state.get("estimated_error_m") or
            nav_state.get("smooth_error_m") or
            nav_state.get("uncertainty_sigma_m") or
            5.2
        )
        estimated_error = round(float(error_val), 2)

        nav_reliability = round((gps_trust * 0.4 + imu_trust * 0.6), 1)

    else:
        # -------------------------------------------------------------
        # GPS HEALTHY / AVAILABLE -> Use existing GPS / Fused GNSS position
        # -------------------------------------------------------------
        position_source = "GPS"
        gps_status_out = "HEALTHY"

        lat = round(float(nav_state.get("latitude", nav_state.get("smooth_latitude", 0.0))), 7)
        lon = round(float(nav_state.get("longitude", nav_state.get("smooth_longitude", 0.0))), 7)

        confidence_val = nav_state.get("confidence") or nav_state.get("ai_confidence_pct") or 96.0
        confidence = round(float(confidence_val), 1)

        error_val = nav_state.get("estimated_error_m") or nav_state.get("uncertainty_sigma_m") or 1.5
        estimated_error = round(float(error_val), 2)

        nav_reliability = round(gps_trust, 1)

    return {
        "latitude": lat,
        "longitude": lon,
        "altitude": altitude_val,
        "speed": speed_val,
        "heading": heading_val,
        "gps_status": gps_status_out,
        "position_source": position_source,
        "confidence": confidence,
        "estimated_error_m": estimated_error,
        "navigation_reliability": nav_reliability,
        "timestamp": t_curr
    }


# CamelCase alias matching user request
getEmergencyLocation = get_emergency_location
